OneSignalSDK/onesignal/src/main/java/com/onesignal/onesignal/location/internal/capture/impl/LocationCapturer.kt

```diff
@@ -0,0 +1,73 @@
+package com.onesignal.onesignal.location.internal.capture.impl
+
+import android.location.Location
+import com.onesignal.onesignal.core.internal.application.IApplicationService
+import com.onesignal.onesignal.core.internal.logging.Logging
+import com.onesignal.onesignal.core.internal.time.ITime
+import com.onesignal.onesignal.core.internal.user.IUserSwitcher
+import com.onesignal.onesignal.location.internal.capture.ILocationCapturer
+import com.onesignal.onesignal.location.internal.common.LocationPoint
+import com.onesignal.onesignal.location.internal.controller.ILocationController
+import com.onesignal.onesignal.location.internal.controller.ILocationUpdatedHandler
+import com.onesignal.onesignal.location.internal.preferences.ILocationPreferencesService
+import java.math.BigDecimal
+import java.math.RoundingMode
+
+internal class LocationCapturer (
+    private val _applicationService: IApplicationService,
+    private val _time: ITime,
+    private val _prefs: ILocationPreferencesService,
+    private val _userSwitcher: IUserSwitcher,
+    private val _controller: ILocationController,
+) : ILocationUpdatedHandler, ILocationCapturer {
+
+    override var locationCoarse: Boolean = false
+
+    init {
+        _controller.subscribe(this)
+    }
+
+    override fun captureLastLocation() {
+        val location = _controller.getLastLocation()
+
+        if(location != null) {
+            capture(location)
+        }
+        else {
+            _prefs.lastLocationTime = _time.currentTimeMillis
+        }
+    }
+
+    override fun onLocationChanged(location: Location) {
+        Logging.debug("LocationController fireCompleteForLocation with location: $location")
+        capture(location)
+    }
+
+    private fun capture(location: Location) {
+        val point = LocationPoint()
+
+        point.accuracy = location.accuracy
+        point.bg = !_applicationService.isInForeground
+        point.type = if (locationCoarse) 0 else 1
+        point.timeStamp = location.time
+
+        // Coarse always gives out 14 digits and has an accuracy 2000.
+        // Always rounding to 7 as this is what fine returns.
+        if (locationCoarse) {
+            point.lat = BigDecimal(location.latitude).setScale(7, RoundingMode.HALF_UP).toDouble()
+            point.log = BigDecimal(location.longitude).setScale(7, RoundingMode.HALF_UP).toDouble()
+        } else {
+            point.lat = location.latitude
+            point.log = location.longitude
+        }
+
+        var userProperties = _userSwitcher.propertiesModel
```

Comment:
I want to try to have all dependencies be against an interface, rather than a concrete class, to have better unit testing control.  `UserManager` implements `IUserManager` which is an external interface, and it implements `IUserSwitcher` which is an internal interface.  My choices were to put it in `IUserManager`, put it in `IUserSwitcher`, or create a new internal interface to put it in.   I chose to put it in `IUserSwitcher` for now, but I suspect there is some outstanding refactoring I will need to do there, in a follow-up PR.